src/lsdb/dask/crossmatch_catalog_data.py

```diff
@@ -0,0 +1,227 @@
+        left_order,
+        left_pixel,
+        right_order,
+        right_pixel,
+        left_hc_structure,
+        right_hc_structure,
+        suffixes,
+    ).crossmatch(**kwargs)
+
+
+# pylint: disable=too-many-locals
+def crossmatch_catalog_data(
+    left: Catalog,
+    right: Catalog,
+    suffixes: Tuple[str, str],
+    algorithm: Type[AbstractCrossmatchAlgorithm]
+    | BuiltInCrossmatchAlgorithm = BuiltInCrossmatchAlgorithm.KD_TREE,
+    **kwargs,
+) -> Tuple[dd.core.DataFrame, DaskDFPixelMap, PixelAlignment]:
+    """Cross-matches the data from two catalogs
+
+    Args:
+        left (Catalog): the left catalog to perform the cross-match on
+        right (Catalog): the right catalog to perform the cross-match on
+        suffixes (Tuple[str,str]): the suffixes to append to the column names from the left and
+            right catalogs respectively
+        algorithm (BuiltInCrossmatchAlgorithm | Callable): The algorithm to use to perform the
+            crossmatch. Can be specified using a string for a built-in algorithm, or a custom
+            method. For more details, see `crossmatch` method in the `Catalog` class.
+        **kwargs: Additional arguments to pass to the cross-match algorithm
+
+    Returns:
+        A tuple of the dask dataframe with the result of the cross-match, the pixel map from HEALPix
+        pixel to partition index within the dataframe, and the PixelAlignment of the two input
+        catalogs.
+    """
+    crossmatch_algorithm = get_crossmatch_algorithm(algorithm)
+
+    # perform alignment on the two catalogs
+    alignment = align_trees(
+        left.hc_structure.pixel_tree,
+        right.hc_structure.pixel_tree,
+        alignment_type=PixelAlignmentType.INNER,
+    )
+    join_pixels = alignment.pixel_mapping
+
+    # align partitions from the catalogs to match the pixel alignment
+    left_aligned_to_join_partitions = align_catalog_to_partitions(
+        left,
+        join_pixels,
+        order_col=PixelAlignment.PRIMARY_ORDER_COLUMN_NAME,
+        pixel_col=PixelAlignment.PRIMARY_PIXEL_COLUMN_NAME,
+    )
+    right_aligned_to_join_partitions = align_catalog_to_partitions(
+        right,
+        join_pixels,
+        order_col=PixelAlignment.JOIN_ORDER_COLUMN_NAME,
+        pixel_col=PixelAlignment.JOIN_PIXEL_COLUMN_NAME,
+    )
+
+    # get lists of HEALPix pixels from alignment to pass to cross-match
+    left_orders = [
+        row[PixelAlignment.PRIMARY_ORDER_COLUMN_NAME]
+        for _, row in join_pixels.iterrows()
+    ]
+    left_pixels = [
+        row[PixelAlignment.PRIMARY_PIXEL_COLUMN_NAME]
+        for _, row in join_pixels.iterrows()
+    ]
+    right_orders = [
+        row[PixelAlignment.JOIN_ORDER_COLUMN_NAME] for _, row in join_pixels.iterrows()
+    ]
+    right_pixels = [
+        row[PixelAlignment.JOIN_PIXEL_COLUMN_NAME] for _, row in join_pixels.iterrows()
+    ]
+
+    # perform the crossmatch on each partition pairing using dask delayed for laziness
```

Comment:
=D

Let's clarify that "laziness" is a good thing here, and it's not laziness on the part of the developers.

Reply:
Lol, yeah good point